index.js

```diff
@@ -67,60 +86,87 @@ let enemyArtistsList = new Map()
 setTable('friendlyTableBody')
 setTable('enemyTableBody')
 
-async function onMessage (event) {
-    let msg = JSON.parse(event.data);
-    if (msg[0] === "pixel")
+/**
+ * Records the placement of a pixel in friendly/enemy artist table.
+ *
+ * @param {number} x - X co-ordinate
+ * @param {number} y - Y co-ordinate.
+ * @param {keyof palette} recvColorIdx - Index number of the color that was painted.
+ * @param {string} _hexTime - Time of pixel placement as a hex string.
+ * Currently unused.
+ * @param {string} address - Wallet address of the person who painted this pixel.
+ * @param {string | null} ens - Wallet ENS, if set.
+ */
+function recordPixelPlacement(x, y, recvColorIdx, _hexTime, address, ens) {
+    const withinImageRange = (
+      img.x <= x && x < img.x + img.w &&
+      img.y <= y && y < img.y + img.h
+    );
+    if (!withinImageRange) {
+      return;
+    }
+
+    const col = drawCtx.getImageData(x, y, 1, 1).data;
+    let expectedColorIdx = approximateColor(col[0], col[1], col[2])
+    let i = (y - img.y) * img.w + x - img.x
+    const wallet = ens ?? address; // use ENS if it exists
```

Comment:
The `??` syntax is called https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator and is a new JavaScript feature.

It was added in Chrome 80 (February 2020), so if people are using MetaMask, their browser should also support this feature.

But we can always replace it with `||` (OR) if think it is worth supporting older browsers (like Internet Explorer) too.

Reply:
I didn't know about this yet, thanks for the info